src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -0,0 +1,919 @@
+        }
+    }
+
+    fn visit_place_base(
+        &mut self,
+        place_base: &PlaceBase<'tcx>,
+        context: PlaceContext,
+        location: Location,
+    ) {
+        trace!(
+            "visit_place_base: place_base={:?} context={:?} location={:?}",
+            place_base,
+            context,
+            location,
+        );
+        self.super_place_base(place_base, context, location);
+
+        match place_base {
+            PlaceBase::Local(_) => {}
+            PlaceBase::Static(box Static{ kind: StaticKind::Promoted(_, _), .. }) => {
+                bug!("Promotion must be run after const validation");
+            }
+
+            PlaceBase::Static(box Static{ kind: StaticKind::Static, def_id, .. }) => {
+                let is_thread_local = self.tcx.has_attr(*def_id, sym::thread_local);
+                if is_thread_local {
+                    self.check_op(ops::ThreadLocalAccess);
+                } else if self.mode == Mode::Static && context.is_mutating_use() {
+                    // this is not strictly necessary as miri will also bail out
+                    // For interior mutability we can't really catch this statically as that
+                    // goes through raw pointers and intermediate temporaries, so miri has
+                    // to catch this anyway
+
+                    self.tcx.sess.span_err(
+                        self.span,
+                        "cannot mutate statics in the initializer of another static",
+                    );
+                } else {
+                    self.check_op(ops::StaticAccess);
+                }
+            }
+        }
+    }
+
+    fn visit_assign(&mut self, dest: &Place<'tcx>, rvalue: &Rvalue<'tcx>, location: Location) {
+        trace!("visit_assign: dest={:?} rvalue={:?} location={:?}", dest, rvalue, location);
+
+        // Error on mutable borrows or shared borrows of values with interior mutability.
+        //
+        // This replicates the logic at the start of `assign` in the old const checker.  Note that
+        // it depends on `HasMutInterior` being set for mutable borrows as well as values with
+        // interior mutability.
+        if let Rvalue::Ref(_, kind, ref borrowed_place) = *rvalue {
+            let rvalue_has_mut_interior = HasMutInterior::in_rvalue(
+                &self.item,
+                self.qualifs.has_mut_interior.get(),
+                rvalue,
+            );
+
+            if rvalue_has_mut_interior {
+                let is_derived_from_illegal_borrow = match *borrowed_place {
+                    // If an unprojected local was borrowed and its value was the result of an
+                    // illegal borrow, suppress this error and mark the result of this borrow as
+                    // illegal as well.
+                    Place { base: PlaceBase::Local(borrowed_local), projection: box [] }
+                        if self.derived_from_illegal_borrow.contains(borrowed_local) => true,
```

Comment:
This is the part of the code which replicates the effect of [clearing `HasMutInterior`](https://github.com/rust-lang/rust/blob/7efe1c6e678a263b9464f2e7f06f552b4d4db5a5/src/librustc_mir/transform/qualify_consts.rs#L777) after an illegal borrow occurs.